hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/AccountStore.java

```diff
@@ -0,0 +1,104 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.service.token.impl;
+
+import static com.hedera.node.app.spi.state.StateKey.ACCOUNT_STORE;
+import static com.hedera.node.app.spi.state.StateKey.ALIASES_STORE;
+import static com.hedera.services.utils.EntityIdUtils.isAlias;
+
+import com.google.protobuf.ByteString;
+import com.hedera.node.app.spi.meta.TransactionMetadata;
+import com.hedera.node.app.spi.meta.impl.SigTransactionMetadata;
+import com.hedera.node.app.spi.state.State;
+import com.hedera.node.app.spi.state.States;
+import com.hedera.services.legacy.core.jproto.JKey;
+import com.hedera.services.state.merkle.MerkleAccount;
+import com.hedera.services.utils.EntityNum;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.Transaction;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+
+/**
+ * Provides methods for interacting with the underlying data storage mechanisms for working with
+ * Accounts.
+ *
+ * <p>This class is not exported from the module. It is an internal implementation detail.
+ */
+public class AccountStore {
+    /** The underlying data storage class that holds the account data. */
+    private final State<EntityNum, MerkleAccount> accountState;
+    /** The underlying data storage class that holds the aliases data built from the state. */
+    private final State<ByteString, EntityNum> aliases;
+
+    /**
+     * Create a new {@link AccountStore} instance.
+     *
+     * @param states The state to use.
+     */
+    public AccountStore(@Nonnull States states) {
+        this.accountState = states.get(ACCOUNT_STORE);
+        this.aliases = states.get(ALIASES_STORE);
+        Objects.requireNonNull(accountState);
+        Objects.requireNonNull(aliases);
+    }
+
+    public TransactionMetadata createAccountSigningMetadata(
+            final Transaction tx,
+            final Optional<JKey> key,
+            final boolean receiverSigReq,
+            final AccountID payer) {
+        if (receiverSigReq && key.isPresent()) {
+            return new SigTransactionMetadata(this, tx, payer, List.of(key.get()));
+        }
+        return new SigTransactionMetadata(this, tx, payer);
+    }
+
+    /**
+     * Returns the account leaf for the given account number. If the account doesn't exist throws
+     * {@link IllegalArgumentException}
+     *
+     * @param id given account number
+     * @return merkle leaf for the given account number
+     */
+    public Optional<MerkleAccount> getAccountLeaf(final AccountID id) {
```

Comment:
We don't want to let any merkle-ness escape from the Store. The Store implementations should completely hide the Merkle stuff, so that no other code in the module has to know or care about anything related to a Merkle tree. There should be an Account type that the other code works with.

Not only does exposing Merkle concepts make testing harder, it also tightly couples the API to a merkle data structure which is unneeded. It also exposes you to risks, because somebody could modify the `MerkleAccount` which **must not happen**. The only thing that modifies the merkle tree is the States when they push to the merkle tree. Otherwise we could end up with all kinds of bugs.

Reply:
For sigRequirements will be using `KeyOrReason` record type to just get the key or return status if there is a failure while fetching the key (i.e., if account doesn't exist or if account is immutable) from `MerkleAccount` and `getAccountLeaf` will not be exposed.